source/extensions/filters/network/thrift_proxy/conn_manager.cc

```diff
@@ -220,6 +222,20 @@ FilterStatus ConnectionManager::ResponseDecoder::messageBegin(MessageMetadataSha
   if (metadata->hasReplyType()) {
     success_ = metadata->replyType() == ReplyType::Success;
   }
+
+  // Is the upstream host draining?
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.thrift_connection_draining")) {
+    metadata_->setDraining(!metadata->headers().get(Headers::get().Drain).empty());
+    metadata->headers().remove(Headers::get().Drain);
+  }
+
+  // Is this host draining?
+  ConnectionManager& cm = parent_.parent_;
+  if (cm.drain_decision_.drainClose()) {
+    // TODO(rgs1): should the key value contain something useful? E.g.: minutes til drain is over?
+    metadata->headers().addReferenceKey(Headers::get().Drain, "true");
```

Comment:
is this the right place to do this? Should it be in ConnectionManager::ResponseDecoder::transportEnd() before transport_->encodeFrame() ? Its strange to manipulate headers in messageBegin when headers are a transport "thing", and not a message thing.

Reply:
It should be in messageBegin, and not transportBegin, so that we can support the header in TTwitter protocol, which reads/adds response headers to metadata in https://github.com/envoyproxy/envoy/blob/eb48804a9e62de3d6770070ef6cd2f70b0727a71/source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc#L974 when reading the response from upstream. Therefore detecting a close should go in messageBegin.

Similarly when on upstream side, the response is buffered until transportEnd. Therefore metadata should be set before the encodeFrame() call. Also it should be set at or after the messageBegin call so that the header is added after all upstream headers passed (due to messageBegin possible getting headers not in transportBegin).

Reply:
Added the last comment inline, for posterity.